optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/bi/DroolsGroupingBiConstraintStream.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.score.stream.drools.bi;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+
+import org.drools.model.Global;
+import org.drools.model.PatternDSL;
+import org.drools.model.Rule;
+import org.drools.model.RuleItemBuilder;
+import org.optaplanner.core.api.score.holder.AbstractScoreHolder;
+import org.optaplanner.core.api.score.stream.uni.UniConstraintCollector;
+import org.optaplanner.core.impl.score.stream.drools.DroolsConstraint;
+import org.optaplanner.core.impl.score.stream.drools.DroolsConstraintFactory;
+import org.optaplanner.core.impl.score.stream.drools.common.DroolsAbstractConstraintStream;
+import org.optaplanner.core.impl.score.stream.drools.common.DroolsLogicalTuple;
+import org.optaplanner.core.impl.score.stream.drools.uni.DroolsAbstractUniConstraintStream;
+
+import static org.drools.model.DSL.declarationOf;
+import static org.drools.model.PatternDSL.pattern;
+
+public class DroolsGroupingBiConstraintStream<Solution_, A, NewA, ResultContainer_, NewB>
+        extends DroolsAbstractBiConstraintStream<Solution_, NewA, NewB> {
+
+    private final DroolsAbstractUniConstraintStream<Solution_, A> parent;
+    private final Function<A, NewA> groupKeyMapping;
+    private final UniConstraintCollector<A, ResultContainer_, NewB> collector;
+    private final AtomicInteger ruleId = new AtomicInteger(-1);
```

Comment:
AtomicInteger has the method `getAndIncrement()` too.
Usually other optaplanner-core code starts from 0 and uses `getAndIncrement()` and this would be called nextRuleId instead.

But there's more complexity in play here, let me grok that first...

Reply:
Ok, I get it.
Instead of using a "final AtomicInteger", can we simply not use a non-final "int"?

AtomicInteger are needed for thread-safety, which will never be in play here (score directors are isolated per thread). AtomicIntegers are sometimes used as a unique id dispenser, which I believed what was going on here.
But it's not: it's actually just a field containing the ruleId. The only reason why the AtomicInteger wrapping is in play, as far as I can tell, is so that the field can be made final?